File: 4026-maximum-gap-between-stations/4026-maximum-gap-between-stations.cpp
class Solution {
public:
    int maximumGap(string skill, string station) {
        int n = skill.size();
        int m = station.size();

        if(n == 1 || m == 1) return 0;
        if(n == m) return 1;

        vector<int> left(n);
        vector<int> right(n);
        int j = 0;

        for(int i = 0; i < n; i++){
            while(skill[i] != station[j]) j++;

            left[i] = j;
            j++;
        }

        j = m-1;

        for(int i = n-1; i >= 0; i--){
            while(skill[i] != station[j]) j--;

            right[i] = j;
            j--;
        }

        int ans = INT_MIN;

        for(int i = 0; i < n-1; i++){
            ans = max(ans, right[i+1] - left[i]);
        }

        return ans;
    }
};

File: 4026-maximum-gap-between-stations/4026-maximum-gap-between-stations.py
class Solution:
    def maximumGap(self, skill: str, station: str) -> int:
        n = len(skill)
        m = len(station)

        if n == 1 or m == 1:
            return 0

        if n == m:
            return 1

        left = [0] * n
        right = [0] * n

        j = 0

        for i in range(n):
            while skill[i] != station[j]:
                j += 1

            left[i] = j
            j += 1

        j = m - 1

        for i in range(n - 1, -1, -1):
            while skill[i] != station[j]:
                j -= 1

            right[i] = j
            j -= 1

        ans = float('-inf')

        for i in range(n - 1):
            ans = max(ans, right[i + 1] - left[i])

        return ans